File: NetDeck/CardFilterPopover.m
//
//  CardFilterPopover.m
//  Net Deck
//
//  Created by Gereon Steffens on 12.01.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "CardFilterPopover.h"
#import "CGRectUtils.h"
#import "CardFilterViewController.h"

@interface CardFilterPopover ()

@property NSArray* sections;
@property NSArray* values;
@property UIButton* button;
@property NSString* type;
@property UIViewController<FilterCallback>* headerView;
@property NSMutableSet* selectedValues;
@property NSMutableArray* sectionToggles;
@property NSMutableArray* collapsedSections;

@property int sectionCount; // number of non-empty section headers
@property NSUInteger totalEntries; // total number of selectable entries

@end

@implementation CardFilterPopover

static UIPopoverController* popover;

+(void) showFromButton:(UIButton *)button inView:(UIViewController<FilterCallback>*)vc entries:(TableData*)entries type:(NSString *)type selected:(id)preselected
{
    CardFilterPopover* filter = [[CardFilterPopover alloc] initWithNibName:@"CardFilterPopover" bundle:nil];
    filter.sections = entries.sections;
    filter.values = entries.values;
    filter.collapsedSections = entries.collapsedSections.mutableCopy;
    filter.button = button;
    filter.type = type;
    filter.headerView = vc;

    if ([preselected isKindOfClass:[NSSet class]])
    {
        filter.selectedValues = [[NSSet setWithSet:preselected] mutableCopy];
    }
    else
    {
        filter.selectedValues = [NSMutableSet set];
        if (preselected && ![preselected isEqualToString:kANY])
        {
            [filter.selectedValues addObject:preselected];
        }
    }
    filter.sectionToggles = [NSMutableArray array];
    
    filter.sectionCount = 0;
    for (NSString* s in entries.sections)
    {
        [filter.sectionToggles addObject:@(NO)];
        [filter.collapsedSections addObject:@(NO)];
        if (s.length > 0)
        {
            ++filter.sectionCount;
        }
    }
    
    filter.totalEntries = 0;
    for (NSArray* arr in entries.values)
    {
        filter.totalEntries += arr.count;
    }
    
    popover = [[UIPopoverController alloc] initWithContentViewController:filter];
    popover.backgroundColor = [UIColor whiteColor];
    
    // make the popover height match the height of the inner tableView
    CGSize tableSize = filter.tableView.frame.size;
    popover.popoverContentSize = tableSize;
    
    CGRect rect = button.frame;
    rect = [button.superview convertRect:rect toView:vc.view];
    
    [popover presentPopoverFromRect:rect inView:vc.view permittedArrowDirections:UIPopoverArrowDirectionLeft animated:NO];
}

+(void) dismiss
{
    [popover dismissPopoverAnimated:NO];
    popover = nil;
}

-(void) dealloc
{
    self.tableView.delegate = nil;
    self.tableView.dataSource = nil;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    
    UITapGestureRecognizer* tableTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(doubleTap:)];
    tableTap.numberOfTapsRequired = 2;
    [self.tableView addGestureRecognizer:tableTap];
    
    [self setTableHeight];
    
    popover.popoverContentSize = self.tableView.frame.size;
}

-(void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self setTableHeight];
}

-(void) doubleTap:(UIGestureRecognizer*) gesture
{
    [CardFilterPopover dismiss];
}

#define CELL_HEIGHT     40
#define HEADER_HEIGHT   25

-(void) setTableHeight
{
    int h = 0;
    for (NSArray* arr in self.values)
    {
        h += CELL_HEIGHT * arr.count;
    }
    h += HEADER_HEIGHT * self.sectionCount;

    self.tableView.scrollEnabled = h > 700;
    h = MIN(h, 700);

    self.tableView.frame = CGRectSetHeight(self.tableView.frame, h);
}

#pragma mark tableview

-(NSInteger) numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.sections.count;
}

-(NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSArray* arr = self.values[section];
    BOOL collapsed = [self.collapsedSections[section] boolValue];
    return collapsed ? 0 : arr.count;
}

-(CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return CELL_HEIGHT;
}

-(CGFloat) tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    NSString* s = self.sections[section];
    return s.length > 0 ? HEADER_HEIGHT : 0;
}

-(NSString*) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    NSString* s = self.sections[section];
    return s.length > 0 ? s : nil;
}

-(UIView*) tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{
    UIView* view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 200, HEADER_HEIGHT)];
    view.backgroundColor = [UIColor colorWithWhite:.9 alpha:1];
    view.tag = section;
    view.userInteractionEnabled = YES;
    
    CGFloat xOffset = self.collapsedSections == nil ? 15 : 25;
    UILabel* label = [[UILabel alloc] initWithFrame:CGRectMake(xOffset, 0, 200, HEADER_HEIGHT)];
    label.font = [UIFont boldSystemFontOfSize:15];
    label.text = self.sections[section];
    
    [view addSubview:label];

    UITapGestureRecognizer* tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didSelectSection:)];
    [view addGestureRecognizer:tap];
    
    UIButton* collapseButton = [UIButton buttonWithType:UIButtonTypeSystem];
    collapseButton.frame = CGRectMake(0, 0, 30, HEADER_HEIGHT);
    collapseButton.tag = section;
    collapseButton.titleLabel.font = [UIFont boldSystemFontOfSize:19];
    [collapseButton addTarget:self action:@selector(collapseSection:) forControlEvents:UIControlEventTouchUpInside];
    
    if (self.collapsedSections)
    {
        BOOL collapsed = [self.collapsedSections[section] boolValue];
        [UIView performWithoutAnimation:^{
            [collapseButton setTitle:collapsed ? @"▹" : @"▿" forState:UIControlStateNormal];
        }];
    }
    
    [view addSubview:collapseButton];
    
    return view;
}

-(UITableViewCell*) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString* cellIdentifier = @"popupCell";
    
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
    }
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    cell.accessoryType = UITableViewCellAccessoryNone;
    
    NSString* value = [self.values objectAtIndexPath:indexPath];
    
    if ([value isEqualToString:kANY])
    {
        cell.textLabel.text = l10n(value);
    }
    else
    {
        cell.textLabel.text = value;
    }
    cell.textLabel.font = [UIFont systemFontOfSize:15];
    cell.textLabel.textColor = [UIColor colorWithRed:0 green:0.5 blue:1 alpha:1];
    
    if ([self.selectedValues containsObject:value] && ![value isEqualToString:kANY])
    {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    }
    
    return cell;
}

-(void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString* value = [self.values objectAtIndexPath:indexPath];
    
    // first ("Any") cell tapped?
    BOOL anyCell = indexPath.row == 0 && indexPath.section == 0;
    
    // if not, and we adding a new selection, and we're 1 shy of checking all possible values, treat as a tap on "Any"
    if (!anyCell && ![self.selectedValues containsObject:value] && self.selectedValues.count == self.totalEntries - 2)
    {
        anyCell = YES;
        value = kANY;
    }

    if (anyCell)
    {
        NSString* title = [NSString stringWithFormat:@"%@: %@", l10n(self.type), l10n(value) ];
        [self.button setTitle:title forState:UIControlStateNormal];
        
        [self.headerView filterCallback:self.button type:[self.type lowercaseString] value:value];
        
        [CardFilterPopover dismiss];
    }
    else
    {
        NSAssert(self.selectedValues != nil, @"nil selectedValue");
        UITableViewCell* cell = [tableView cellForRowAtIndexPath:indexPath];

        if ([self.selectedValues containsObject:value])
        {
            cell.accessoryType = UITableViewCellAccessoryNone;
            [self.selectedValues removeObject:value];
        }
        else
        {
            cell.accessoryType = UITableViewCellAccessoryCheckmark;
            [self.selectedValues addObject:value];
        }
        
        [self filterWithMultipleSelection];
    }
}

-(void) collapseSection:(UIButton*) sender
{
    NSAssert(self.collapsedSections != nil, @"collapsedSections is nil");
    
    BOOL collapsed = [self.collapsedSections[sender.tag] boolValue];
    // NSLog(@"collapse section %d = %d", sender.tag, collapsed);
    
    [UIView performWithoutAnimation:^{
        [sender setTitle:collapsed ? @"▹" : @"▿" forState:UIControlStateNormal];
    }];
    
    self.collapsedSections[sender.tag] = @(!collapsed);
    
    [self.tableView reloadData];
}

-(void) didSelectSection:(UIGestureRecognizer*)gesture
{
    NSInteger section = gesture.view.tag;
    BOOL collapsed = [self.collapsedSections[section] boolValue];
    if (collapsed)
    {
        return;
    }
    
    BOOL on = [[self.sectionToggles objectAtIndex:section] boolValue];
    // NSString* s = self.sections[section];
    // NSLog(@"section toggle %@ %d", s, on);
    self.sectionToggles[section] = @(!on);
    
    if (!on)
    {
        [self.selectedValues addObjectsFromArray:self.values[section]];
    }
    else
    {
        for (NSString* s in self.values[section])
        {
            [self.selectedValues removeObject:s];
        }
    }
    
    // NSLog(@"selected: %@", self.selectedValues);
    
    [self.tableView reloadData];
    [self filterWithMultipleSelection];
}

-(void) filterWithMultipleSelection
{
    // NSLog(@"notify %@ %@", self.type, self.selectedValues);
    NSString* selected = self.selectedValues.count == 0 ? l10n(kANY) : (self.selectedValues.count == 1 ? [[self.selectedValues allObjects] objectAtIndex:0] : @"⋯");
    NSString* title = [NSString stringWithFormat:@"%@: %@", l10n(self.type), selected];
    [self.button setTitle:title forState:UIControlStateNormal];
    
    [self.headerView filterCallback:self.button type:[self.type lowercaseString] value:self.selectedValues];
}
@end

File: NetDeck/CardFilterPopover.swift
//
//  CardFilterPopover.swift
//  NetDeck
//
//  Created by Gereon Steffens on 06.01.17.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import UIKit

protocol FilteringViewController: class {
    
    func filterCallback(attribute: FilterAttribute, value: FilterValue)
    
    var view: UIView! { get }
    
    func present(_ : UIViewController, animated: Bool, completion: (()->Void)?)
}

class CardFilterPopover: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    @IBOutlet weak var tableView: UITableView!
    
    private var sections = [String]()
    private var values = [[String]]()
    private var button: UIButton!
    private var attribute: FilterAttribute!
    private weak var filteringViewController: FilteringViewController!
    private var selectedValues = Set<String>()
    private var sectionToggles = [Bool]()
    private var collapsedSections: [Bool]?
    
    private var sectionCount = 0
    private var totalEntries = 0
    
    private static var popover: CardFilterPopover!
    
    static func showFrom(button: UIButton, inView vc: FilteringViewController, entries: TableData<String>, attribute: FilterAttribute, selected: FilterValue?) {
        popover = CardFilterPopover()
        popover.sections = entries.sections
        popover.values = entries.values
        popover.collapsedSections = entries.collapsedSections
        popover.button = button
        popover.attribute = attribute
        popover.filteringViewController = vc
        
        popover.selectedValues.removeAll()
        if let selectedSet = selected?.strings, selectedSet.count > 1 {
            popover.selectedValues = selectedSet
        } else if let selectedStr = selected?.string, selectedStr != Constant.kANY {
            popover.selectedValues = Set([selectedStr])
        }
        
        popover.sectionToggles.removeAll()
        
        popover.sectionCount = 0
        for s in entries.sections {
            popover.sectionToggles.append(false)
            popover.collapsedSections?.append(false)
            if s.count > 0 {
                popover.sectionCount += 1
            }
        }
        
        popover.totalEntries = 0
        for arr in popover.values {
            popover.totalEntries += arr.count
        }
        
        let rect = button.superview?.convert(button.frame, to: vc.view) ?? CGRect.zero
        
        popover.modalPresentationStyle = .popover
        popover.popoverPresentationController?.sourceRect = rect
        popover.popoverPresentationController?.sourceView = vc.view
        popover.popoverPresentationController?.permittedArrowDirections = .left
        
        vc.present(popover, animated: false, completion: nil)
    }
    
    static func dismiss() {
        popover?.dismiss(animated: false, completion: nil)
        popover = nil
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.tableFooterView = UIView(frame: CGRect.zero)
        
        let tableTap = UITapGestureRecognizer(target: self, action: #selector(doubleTap(_:)))
        tableTap.numberOfTapsRequired = 2
        self.tableView.addGestureRecognizer(tableTap)
        
        self.setTableHeight()
        
        self.preferredContentSize = self.tableView.frame.size
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.setTableHeight()
    }
    
    @objc func doubleTap(_ gesture: UITapGestureRecognizer) {
        if gesture.state == .ended {
            CardFilterPopover.dismiss()
        }
    }
    
    private let cellHeight = 40
    private let headerHeight = 25
    private let tableWidth = 220
    
    func setTableHeight() {
        var h = 0
        for arr in self.values {
            h += cellHeight * arr.count
        }
        h += headerHeight * self.sectionCount
        
        self.tableView.isScrollEnabled = h > 700
        h = min(h, 700)
        
        var frame = self.tableView.frame
        frame.size.height = CGFloat(h)
        self.tableView.frame = frame
    }
    
    // MARK: - table view
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.sections.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let arr = self.values[section]
        let collapsed = self.collapsedSections?[section] ?? false
        return collapsed ? 0 : arr.count
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return CGFloat(cellHeight)
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        let s = self.sections[section]
        return s.count > 0 ? CGFloat(headerHeight) : 0
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let s = self.sections[section]
        return s.count > 0 ? s : nil
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let view = UIView(frame: CGRect(x: 0, y: 0, width: tableWidth, height: headerHeight))
        view.backgroundColor = .secondarySystemBackground
        view.tag = section
        view.isUserInteractionEnabled = true

        let hackOffset: Int = {
            if #available(iOS 13.0, *) { return 10 } else { return 0 }
        }()

        let xOffset = hackOffset + (self.collapsedSections == nil ? 15 : 25)
        let label = UILabel(frame: CGRect(x: xOffset, y:0, width: tableWidth, height: headerHeight))
        label.font = UIFont.boldSystemFont(ofSize: 15)
        label.text = self.sections[section]
        view.addSubview(label)
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(self.didSelectSection(_:)))
        view.addGestureRecognizer(tap)
        
        if self.collapsedSections != nil {
            let collapseButton = UIButton(type: .system)
            collapseButton.frame = CGRect(x: hackOffset, y: 0, width: 30, height: headerHeight)
            collapseButton.tag = section
            collapseButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 19)
            collapseButton.addTarget(self, action: #selector(self.collapseSection(_:)), for: .touchUpInside)
            view.addSubview(collapseButton)
        
            let collapsed = self.collapsedSections![section]
            UIView.performWithoutAnimation {
                collapseButton.setTitle(collapsed ? "▹" : "▿", for: .normal)
            }
        }
        return view
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "popupCell"
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier) ?? {
            let c = UITableViewCell(style: .default, reuseIdentifier: identifier)
            c.selectionStyle = .none
            c.textLabel?.font = UIFont.systemFont(ofSize: 15)
            c.textLabel?.textColor = UIColor(red: 0, green: 0.5, blue: 1, alpha: 1)
            return c
        }()

        cell.accessoryType = .none
        
        let value = self.values[indexPath.section][indexPath.row]
        let any = value == Constant.kANY
        if any {
            cell.textLabel?.text = value.localized()
        } else {
            cell.textLabel?.text = value
        }
        
        if self.selectedValues.contains(value) && !any {
            cell.accessoryType = .checkmark
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        var value = self.values[indexPath.section][indexPath.row]
        
        // is this the first ("Any") cell?
        var anyCell = indexPath.row == 0 && indexPath.section == 0
        
        // if not, and we adding a new selection, and we're 1 shy of checking all possible values, treat as a tap on "Any"
        if !anyCell && !self.selectedValues.contains(value) && self.selectedValues.count == self.totalEntries - 2 {
            anyCell = true
            value = Constant.kANY
        }
        
        if anyCell {
            let title = self.attribute.localized() + ": " + value.localized()
            self.button.setTitle(title, for: .normal)
            
            self.filteringViewController.filterCallback(attribute: self.attribute, value: FilterValue.string(value))
            
            CardFilterPopover.dismiss()
        } else {
            let cell = tableView.cellForRow(at: indexPath)
            if self.selectedValues.contains(value) {
                cell?.accessoryType = .none
                self.selectedValues.remove(value)
            } else {
                cell?.accessoryType = .checkmark
                self.selectedValues.insert(value)
            }
            
            self.filterWithMultipleSelection()
        }
    }
    
    func filterWithMultipleSelection() {
        let selected: String
        let value: FilterValue
        switch selectedValues.count {
        case 0:
            selected = Constant.kANY.localized()
            value = FilterValue.string(Constant.kANY)
        case 1:
            let str = Array(self.selectedValues)[0]
            selected = str
            value = FilterValue.string(str)
        default:
            selected = "⋯"
            value = FilterValue.strings(selectedValues)
        }
        
        let title = self.attribute.localized() + ": " + selected
        
        self.button.setTitle(title, for: .normal)
        
        self.filteringViewController.filterCallback(attribute: self.attribute, value: value)
    }
    
    @objc func collapseSection(_ sender: UIButton) {
        assert(self.collapsedSections != nil)
        
        let collapsed = self.collapsedSections![sender.tag]
        
        UIView.performWithoutAnimation {
            sender.setTitle(collapsed ? "▹" : "▿", for: .normal)
        }
        self.collapsedSections![sender.tag] = !collapsed
        self.tableView.reloadData()
    }
    
    @objc func didSelectSection(_ gesture: UITapGestureRecognizer) {
        guard let section = gesture.view?.tag else {
            return
        }
        let collapsed = self.collapsedSections?[section] ?? false
        if collapsed {
            return
        }
        
        let on = self.sectionToggles[section]
        self.sectionToggles[section] = !on

        if !on {
            self.selectedValues.formUnion(self.values[section])
        } else {
            self.selectedValues.subtract(self.values[section])
        }
        
        self.tableView.reloadData()
        self.filterWithMultipleSelection()
    }
    
}
